gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/ApplicationFactorSettingsUpgrader.java

```diff
@@ -0,0 +1,191 @@
+                    return Single.just(false);
+                });
+    }
+
+    private static MFASettings migrateEnrollSettings(Application app) {
+        final var mfaSettings = app.getSettings().getMfa();
+        final var enrollSettings = new EnrollSettings();
+        mfaSettings.setEnroll(enrollSettings);
+        enrollSettings.setActive(true);
+        if (!mfaSettings.getEnrollment().getForceEnrollment()) {
+            enrollSettings.setType(MfaEnrollType.OPTIONAL);
+            enrollSettings.setForceEnrollment(false);
+            enrollSettings.setSkipTimeSeconds(mfaSettings.getEnrollment().getSkipTimeSeconds());
+        } else {
+            enrollSettings.setType(MfaEnrollType.REQUIRED);
+            enrollSettings.setForceEnrollment(true);
+        }
+        return mfaSettings;
+    }
+
+    private static void migrateChallengeSettings(Application app) {
+        final var mfaSettings = app.getSettings().getMfa();
+        final var challengeSettings = new ChallengeSettings();
+        mfaSettings.setChallenge(challengeSettings);
+        challengeSettings.setActive(true);
+        if (app.getSettings().getRiskAssessment() != null && app.getSettings().getRiskAssessment().isEnabled() && hasLength(mfaSettings.getAdaptiveAuthenticationRule())) {
+            challengeSettings.setType(MfaChallengeType.RISK_BASED);
+            challengeSettings.setChallengeRule(mfaSettings.getAdaptiveAuthenticationRule());
+        } else if (hasLength(mfaSettings.getAdaptiveAuthenticationRule())) {
+            challengeSettings.setType(MfaChallengeType.CONDITIONAL);
+            challengeSettings.setChallengeRule(mfaSettings.getAdaptiveAuthenticationRule());
+        } else {
+            challengeSettings.setType(MfaChallengeType.REQUIRED);
+        }
+    }
+
+    private static void migrateStepUpSettings(Application app) {
+        final var mfaSettings = app.getSettings().getMfa();
+        if (hasLength(mfaSettings.getStepUpAuthenticationRule())) {
+            final var stepUpSettings = new StepUpAuthenticationSettings();
+            stepUpSettings.setActive(true);
+            stepUpSettings.setStepUpAuthenticationRule(mfaSettings.getStepUpAuthenticationRule());
+            mfaSettings.setStepUpAuthentication(stepUpSettings);
+        }
+    }
+
+    private static void moveFactorIdsIntoFactorSettings(Application app, List<String> recoveryCodeIds) {
+        final var factorSettings = new HashSet<ApplicationFactorSettings>();
+        boolean defaultFactor = false;
+        for (var id : app.getFactors()) {
+            if (!defaultFactor) {
+                // default factor need to be the first factor from the list
+                // but recovery code factor shouldn't be a default candidate
+                defaultFactor = !recoveryCodeIds.contains(id);
+            }
+            factorSettings.add(new ApplicationFactorSettings(id, null, defaultFactor));
```

Comment:
application factor settings has been changed a few days ago:

this is the new version. It is used inside FactorMongoSettings
@Getter
@Setter
@NoArgsConstructor
public class ApplicationFactorSettings {
    private String id;
    private String selectionRule;

    public ApplicationFactorSettings(ApplicationFactorSettings applicationFactorSettings) {
        this.id = applicationFactorSettings.id;
        this.selectionRule = applicationFactorSettings.selectionRule;
    }
}


Please use latest destination branch code.